mtx/scsieject.1
.\" scsieject.1  Document Copyright 2007 Robert Nelson
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to the Free
.\" Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139,
.\" USA.
.\"
.TH scsieject 1 scsieject1.0
.SH NAME
scsieject \- control SCSI tape devices
.SH SYNOPSIS
scsieject [-f <scsi-generic-device>] commands
.SH DESCRIPTION
The
.B scsieject
command controls SCSI devices in a platform-independent
manner. As long as 'mtx' works on the platform, so does 'scsieject'.
.SH OPTIONS
The first argument, given following
.B -f
, is the SCSI generic device corresponding to your tape drive.
Consult your operating system's documentation for more information (for
example, under Linux these are generally /dev/sg0 through /dev/sgXX,
under FreeBSD these are /dev/pass0 through /dev/passX. Under Solaris
this is usually the same as your tape drive (Solaris has a SCSI passthrough
ioctl). You can set the STAPE or TAPE environment variable rather
than use -f.
.P
.SH COMMANDS
.TP 10
.B load
Load the medium into the drive.  When this command is issued to a CD/DVD drive
and the tray is extended the tray will be retracted if the drive is capable of it.

.TP 10
.B unload
Unload the medium from the drive (also known as eject).  When this command is issued
to a CD/DVD drive or a tape drive the media will be ejected if the device supports it.

.TP 10
.B start
Start the device.  Some devices require a start command after a media changer has
loaded new media into the device.

.TP 10
.B stop
Stop the device.  Some devices require a stop command prior to unloading the medium
from the device when using a media changer.

.TP 10
.B lock
Lock the device.  Locks the device so that the medium cannot be removed manually.

.TP 10
.B unlock
Unlock the device.  Unlocks the device so that the medium can be removed manually.

.SH AUTHORS
This program was written by Robert Nelson <robertnelson@users.sourceforge.net>
based on the scsitape program written by Eric Lee Green <eric@badtux.org>.
Major portions of the 'mtxl.c' library used herein were written by
Leonard Zubkoff.
.P

.SH HINTS
Under Linux,
.B lsscsi \-\-generic
will tell you what SCSI devices you have and their device names.
.P
Under FreeBSD,
.B camcontrol devlist
will tell you what SCSI devices you
have, along with which
.B pass
device controls them.
.P
Under Solaris 7 and 8,
.B /usr/sbin/devfsadm -C
will clean up your /devices directory. Then
.B find /devices -name 'st@*' -print
will return a list of all tape drives. /dev on Solaris is apparently only
of historical interest.

.SH BUGS AND LIMITATIONS
There are no known bugs or limitations.

.SH AVAILABILITY
This version of
.B scsieject
is currently being maintained by Robert Nelson <robertnelson@users.sourceforge.net>
as part of the 'mtx' suite of programs. The 'mtx' home page is
http://mtx.sourceforge.net and the actual code is currently available there and via
SVN from http://sourceforge.net/projects/mtx.

.SH SEE ALSO
.BR loaderinfo (1), tapeinfo (1), mtx (1)


mtx/CHANGES
CHANGES, mtx 1.3.11
- loaderinfo: Print RequestSense information if an operation fails.
- loaderinfo: Print all information from the Device Configuration Page.
- mtx: Fix problems with a lot of loaders determining the number of elements.
- mtx: Implement Previous command.
- mtx: Fix bugs in Last command.
- scsieject: New tool that handles start/stop, load/unload.
- tapeinfo: Fix inconsistencies in output.
- Fix timeout handling on Linux.
- Merge in fixes from the Debian and FreeBSD distributions.

CHANGES, mtx 1.3.10
- Add support for Microsoft Windows.
- Add support for Sony VGP-XL1B(2) Media Changer.
  Thanks to Will (nodenet at hotmail dot com)
- Add build support for Microsoft Windows using MinGW native and Linux
  cross-compile.
- Add build support for Microsoft Windows using Microsoft Visual Studio 2005.
- Add support for building on x86_64.
- Add more debugging information.
- Eliminate compiler warnings.

CHANGES, mtx 1.3.9
- Cleaned up scsi_linux.c a little to eliminate around 40 lines of code. 
- Change to using SG_IO ioctl rather than write-read interface. This
  should make us a little safer, as well as (on 2.6.x) letting us issue SCSI 
  commands to regular block devices as well as to /dev/sgXX devices. 
  WARNING: Can cause the system to *CRASH* if the SCSI device is in use, due 
  to brokenness inside the Linux kernel! It's always preferable to use the 
  /dev/sgXX device, which has specialness in its buffer handling that bypasses
  some of the brokenness elsewhere in the SCSI subsystem.
- Check SG version before using SG_IO interface, so that if we're run on a
  Linux 2.2 system and we were compiled on Linux 2.4 or above, 
- If a tape is in the drive, make tapeinfo print out its partition info. Seems
  to work on my DAT drive, anyhow (shrug). 
- Minor cleanups in mtxl.c (other cleanups necessary in tapeinfo and etc.). 

CHANGES, mtx 1.3.8
- Forgot about 8 bytes for header
- Handle pedantic loader whose firmware writer spends too much time
  parsing phrases in the SCSI spec and not enough time in the
  real world. 

CHANGES, mtx 1.3.6
- The You Gotta Be F'ing Kidding Release (see rant on mailing list)
- Added AIX support with http://fz.eryx.net/aix/ GSC driver, courtesy
  of Steve Heck. 
- Make __WEIRD_CHAR_SUPPRESS the default to stop barcode garbling
- Fix core dump on invalid inputs in mtx.c
- Add retry with bigger ALLOCATION_LENGTH if 
  BYTE_COUNT_OF_REPORT_AVAILABLE is > than our original estimate. (see rant
  on mailing list). 
- Shut up the low-level SCSI sg_read and sg_write routines, which were poofting
  all over the place on innocuous things like no EAAP. 

CHANGES, mtx 1.3.5
- The I-Need-A-Job Release (see http://badtux.org) :-}
- Change Makefile so that it works w/systems whose 'install' program
  accepts only one argument. That's it. 
- Uncomment the Exabyte hack in the drive parsing code so that it works
  with ATL loaders again. 
- Add hack for ATL stacked loaders to keep them from reporting ghost slots

CHANGES, mtx 1.3.4
- Forward-ported the ATL/Exabyte patches from mtx 1.2
- Added progress indicator to 'tapeinfo' for DAT drives. (PLEASE TEST)
- Purged a few more estinc.com out of the documentation, point to bru.com.

CHANGES, mtx 1.3.3
- Reversed James' long descriptor patch until can figure out what's wrong with
  it.
- Fixed some debug declarations, courtesy of Kevin Wang. 
- Point COMPATIBILITY file at mtx.sourceforge.net.

CHANGES, mtx 1.3.2
- Merged Makefile changes from mtx-1.2 code
- Changes from James Dugal to get all info for newer loaders, fix debugging

CHANGES, mtx 1.3.1
- add an Exchange command

CHANGES, mtx 1.3.0
- New Linux SCSI subsystem interface for 2.4 kernels, ripped boldly 
  from sgtools by Doug Gilbert
- mtx 'next' now skips blank slots, courtesy of Christopher McCrory
- mtx 'unload' now prints an error message telling you that you need to
  eject the tape in the drive prior to unloading it, under certain
  conditions. 
- Started work on 'nsmhack' for controlling NSM jukeboxes. 
- copy_barcode was off by one. 
- 'position to element' command now added, courtesy of Mahlon Stacy

CHANGES, mtx 1.2.15:
- Some Solaris fixes, courtesy of Matt Ward
- Fix URL in .spec file

CHANGES, mtx 1.2.14:
- Fix so it'll work if 0 is result of SCSI open (e.g., in cron jobs on Linux)
- Move changelog to end of .spec file for easier reading
- Added a bit of text to beginning of COMPATIBILITY file

CHANGES, mtx 1.2.13:
- Fixed some autoloader bugs w/autoloaders that don't report an arm. 
- Fixed barcode backoff.
- Added "nobarcode" option
- Increased timeout for 'mtx inventory' to 30 minutes
  (note: may increase this even more if needed, please let me know!)
- Shortened timeout for 'mtx inquiry' to 30 seconds
- tapeinfo now prints SCSI ID/LUN info if available (only on Linux at the
  moment, sigh). 
- update documentation w/new email addresses, updated compile directions,
  various fixes. 

CHANGES, mtx 1.2.12:
- Fix FreeBSD compile bugs
- Fix SGI compile bugs
- Add HP/UX port (I hope!), courtesy of William Smith. 
- Re-wrote ReadElementStatus to make work for %!@# brain dead firmware that
  reports non-existent drives (I hope!). Also has side-effect of now working
  with multiple-arm libraries (though it only sees first arm!).
- Cleaned up all -Wall messages.
- Cleaned up Linux Sparc, installs loaderinfo.1, courtesy of Matt Dainty.
- tapeinfo now reports status of CheckUnitReady.  
- tapeinfo no longer puts out Block Position if CheckUnitReady says 'no'.
- tapeinfo now puts out Density Code and Medium Type/Not Loaded (modification
    of patches sent in by Bob Rahe)

CHANGES, mtx 1.2.11:
- Added a GNU autoconf Makefile.in (still provide a Makefile for your use)
  *WARNING* autoconf is not yet working on all supported OS's! You may need
  to do the old-fashioned 'edit Makefile' bit!
- Changed mtx.h and mtxl.c to include and define various things based upon
  what features autoconf detected (e.g., if 'camlib.h', use FreeBSD-style

mtx/scsieject.cpp
/* Copyright 2007, Robert Nelson
 *   Released under terms of the GNU General Public License as
 * required by the license on 'mtxl.c'.
 * $Date: 2007-01-28 19:23:33 -0800 (Sun, 28 Jan 2007) $
 * $Revision: 125 $
 */

/* This is a generic SCSI device control program. It operates by
 * directly sending commands to the device.
 */

/* 
 *	Commands:
 *		load -- Load medium
 *		unload -- Unload medium
 *		start -- Start device
 *		stop -- Stop device
 *		lock -- Lock medium
 *		unlock -- Unlock medium
 */

#include <stdio.h>
#include <string.h>

#include "mtx.h"
#include "mtxl.h"

#if HAVE_UNISTD_H
#include <unistd.h>
#endif

#if HAVE_SYS_TYPES_H
#include <sys/types.h>
#endif

#ifdef _MSC_VER
#include <io.h>
#endif

char *argv0;

/* the device handle we're operating upon. */
static char *device;  /* the device name. */
static DEVICE_TYPE DeviceFD = (DEVICE_TYPE) -1;

static int S_load(void);
static int S_unload(void);
static int S_start(void);
static int S_stop(void);
static int S_lock(void);
static int S_unlock(void);

struct command_table_struct
{
	char *name;
	int (*command)(void);
}
	command_table[] =
{
	{ "load", S_load },
	{ "unload", S_unload },
	{ "start", S_start },
	{ "stop", S_stop },
	{ "lock", S_lock },
	{ "unlock", S_unlock },
	{ NULL, NULL } /* terminate list */
};

void Usage(void)
{
	FatalError("Usage: scsieject -f <generic-device> <command> where <command> is:\n load | unload | start | stop | lock | unlock\n");
}

/* open_device() -- set the 'DeviceFD' variable.... */
void open_device(void)
{
	if (DeviceFD != -1)
	{
		SCSI_CloseDevice("Unknown", DeviceFD);
	}

	DeviceFD = SCSI_OpenDevice(device);
}

/* we see if we've got a file open. If not, we open one :-(. Then
 * we execute the actual command. Or not :-(. 
 */ 
int execute_command(struct command_table_struct *command)
{
	/*
	 * If the device is not already open, then open it from the 
	 * environment.
	 */
	if (DeviceFD == -1)
	{
		/* try to get it from STAPE or TAPE environment variable... */
		if ((device = getenv("STAPE")) == NULL &&
			(device = getenv("TAPE")) == NULL)
		{
			Usage();	/* Doesn't return */
		}

		open_device();
	}

	/* okay, now to execute the command... */
	return command->command();
}


/* parse_args():
 * Basically, we are parsing argv/argc. We can have multiple commands
 * on a line, such as "load start" to load a tape and start the device.
 * We execute these commands one at a time as we come to them. If we don't 
 * have a -f at the start and the default device isn't defined in a TAPE or 
 * STAPE environment variable, we exit.
 */ 

int parse_args(int argc, char **argv)
{
	int index, retval;
	struct command_table_struct *command;

	argv0 = argv[0];

	for (index = 1; index < argc; index++)
	{
		if (strcmp(argv[index], "-f") == 0)
		{
			index++;
			if (index >= argc)
			{
				Usage();	/* Doesn't return */
			}
			device = argv[index];
			open_device();
		}
		else
		{
			for (command = &command_table[0]; command->name != NULL; command++)
			{
				if (strcmp(command->name, argv[index]) == 0)
				{
					break;
				}
			}

			if (command->name == NULL)
			{
				Usage();	/* Doesn't return */
			}

			retval = execute_command(command);

			if (retval < 0)
			{
				/* Command failed, we probably shouldn't continue */
				return retval;
			}
		}
	}

	return 0;
}

int S_load(void)
{
	int result = LoadUnload(DeviceFD, 1);

	if (result < 0)
	{
		fputs("scsieject: load failed\n", stderr);
		fflush(stderr);
	}

	return result;
}

int S_unload(void)
{
	int result = LoadUnload(DeviceFD, 0);

	if (result < 0)
	{
		fputs("scsieject: unload failed\n", stderr);
		fflush(stderr);
	}

	return result;
}

int S_start(void)
{
	int result = StartStop(DeviceFD, 1);

	if (result < 0)
	{
		fputs("scsieject: start failed\n", stderr);
		fflush(stderr);
	}

	return result;
}

int S_stop(void)
{
	int result = StartStop(DeviceFD, 0);

	if (result < 0)
	{
		fputs("scsieject: stop failed\n", stderr);
		fflush(stderr);
	}

	return result;
}

int S_lock(void)
{
	int result = LockUnlock(DeviceFD, 1);

	if (result < 0)
	{
		fputs("scsieject: lock failed\n", stderr);
		fflush(stderr);
	}

	return result;
}

int S_unlock(void)
{
	int result = LockUnlock(DeviceFD, 0);

	if (result < 0)
	{
		fputs("scsieject: unlock failed\n", stderr);
		fflush(stderr);
	}

	return result;
}

/* See parse_args for the scoop. parse_args does all. */
int main(int argc, char **argv)
{
	parse_args(argc, argv);

	if (device)
	{
		SCSI_CloseDevice(device, DeviceFD);
	}

	exit(0);
}
